multiwallet.go

```diff
@@ -0,0 +1,403 @@
+}
+
+func (mw *MultiWallet) RestoreWallet(seedMnemonic, privatePassphrase string, privatePassphraseType int32) (*LibWallet, error) {
+	if mw.activeSyncData != nil {
+		return nil, errors.New(ErrSyncAlreadyInProgress)
+	}
+
+	properties := &Wallet{
+		PrivatePassphraseType: privatePassphraseType,
+		DiscoveredAccounts:    false,
+	}
+
+	return mw.createWallet(properties, seedMnemonic, privatePassphrase)
+}
+
+func (mw *MultiWallet) createWallet(w *Wallet, seedMnemonic, privatePassphrase string) (*LibWallet, error) {
+
+	libWallet, err := mw.saveWalletToDatabase(w)
+	if err != nil {
+		return nil, err
+	}
+
+	err = libWallet.CreateWallet(privatePassphrase, seedMnemonic)
+	if err != nil {
+		return nil, err
+	}
+
+	go mw.listenForTransactions(libWallet)
+
+	return libWallet, nil
+}
+
+func (mw *MultiWallet) saveWalletToDatabase(w *Wallet) (*LibWallet, error) {
+	// saving struct to update ID property with an autogenerated value
+	err := mw.db.Save(w)
+	if err != nil {
+		return nil, err
+	}
+
+	// delete from database if not created successfully
+	defer func() {
+		if err != nil {
+			mw.db.DeleteStruct(w)
+		}
+	}()
+
+	walletDataDir := filepath.Join(mw.rootDir, strconv.Itoa(w.ID))
+	os.MkdirAll(walletDataDir, os.ModePerm) // create wallet dir
+
+	w.DataDir = walletDataDir
+	err = mw.db.Save(w) // update database with complete wallet information
+	if err != nil {
+		return nil, err
+	}
+
+	w.walletDbDriver = mw.dbDriver
+	w.netType = mw.activeNet.Name
+	libWallet, err := NewLibWallet(w)
+	if err != nil {
+		return nil, err
+	}
+
+	mw.libWallets[w.ID] = libWallet
+	return libWallet, nil
+}
+
+func (mw *MultiWallet) WalletNameExists(walletName string) (bool, error) {
+
+	if strings.HasPrefix(walletName, "wallet-") {
+		return false, errors.E(ErrReservedWalletName)
+	}
+
+	err := mw.db.One("Name", walletName, &Wallet{})
+	if err == nil {
+		return true, nil
+	} else if err != storm.ErrNotFound {
+		return false, err
+	}
+
+	return false, nil
+}
+
+func (mw *MultiWallet) GetWallet(walletID int) *LibWallet {
+	if lw, exists := mw.libWallets[walletID]; exists {
+		return lw
+	}
+	return nil
+}
+
+func (mw *MultiWallet) OpenWallets(pubPass []byte) error {
+	if mw.activeSyncData != nil {
+		return errors.New(ErrSyncAlreadyInProgress)
+	}
+
+	for _, lw := range mw.libWallets {
+		err := lw.OpenWallet(pubPass)
+		if err != nil {
+			return err
+		}
+
+		go mw.listenForTransactions(lw)
+	}
+
+	return nil
+}
+
+func (mw *MultiWallet) OpenWallet(walletID int, pubPass []byte) error {
+	if mw.activeSyncData != nil {
+		return errors.New(ErrSyncAlreadyInProgress)
+	}
+	libWallet, ok := mw.libWallets[walletID]
+	if ok {
+		err := libWallet.OpenWallet(pubPass)
+		if err != nil {
+			return err
+		}
+
+		go mw.listenForTransactions(libWallet)
+		return nil
+	}
+
+	return errors.New(ErrNotExist)
+}
+
+func (mw *MultiWallet) UnlockWallet(walletID int, privPass []byte) error {
+	w, ok := mw.libWallets[walletID]
+	if ok {
+		return w.UnlockWallet(privPass)
```

Comment:
As a continuation of this https://github.com/raedahgroup/dcrlibwallet/pull/57#discussion_r344601645, is the wallet re-locked after discovery is completed?

Reply:
https://github.com/raedahgroup/dcrlibwallet/blob/260fc17b1f08c86ae2546bddf229b9d4dfbde9e1/syncnotification.go#L336